### PR TITLE
Issue 589 - Variadic function for size zero tests

### PR DESCRIPTION
**Pull request by andrjohns**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Addressing issue #589, adds function to check whether inputs are of length 0. Used to reduce code duplication in probability functions

#### Intended Effect:
Simplify code

#### How to Verify:
```
./runTests.py test/unit/math/prim/mat/fun/size_zero_test.cpp
```
#### Side Effects:
Cpplint doesn't like the use of Rvalue references in the function, returns "RValue references are an unapproved C++ feature."
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Andrew Johnson

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
I can review this one.

**Comment by syclik** [comment]
Jenkins, ok to test.

File: stan/math/prim/mat/fun/size_zero.hpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_SIZE_ZERO_HPP
+#define STAN_MATH_PRIM_MAT_FUN_SIZE_ZERO_HPP
+
+#include <stan/math/prim/mat/meta/length.hpp>
+
```
Comment (syclik) [inline stan/math/prim/mat/fun/size_zero.hpp]: Please include the header for `std::forward`. I think it's in `<utility>`.

File: stan/math/prim/mat/fun/size_zero.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+     *
+     * @param x arguments
+     * @return 0 or 1
+     */
+    template <typename T>
+    inline bool size_zero(T& x) {
+      return !length(x);
+    }
+
+    template <typename T, typename ... Ts>
```
Comment (syclik) [inline stan/math/prim/mat/fun/size_zero.hpp]: Copy doc over to this function (with the `xs` argument added to the doc).

**Comment by andrjohns** [comment]
The only issue with this one is that cpplint will cause the build to fail because ```RValue references are an unapproved C++ feature.  [build/c++11] [3]```

To get around this, can either: 
- stop cpplint from checking ```[build/c++11]``` errors altogether
- Remove the Rvalue references check from cpplint.py
- Trick cpplint into thinking that it's not an rvalue reference

I think the third option would be best, simply because it's the least likely to affect any other code. To make cpplint think that it's not an Rvalue reference, simply need to change 

```bool size_zero(T& x, Ts&& ... xs)```

to

```bool size_zero(T& x, Ts && ... xs)```

(i.e. add a space between ```Ts``` and ```&&``` so that cpplint thinks that it's a boolean expression)

**Comment by syclik** [comment]
My inclination would be to ignore build/c++11 warnings because we're now using C++11!

**Comment by andrjohns** [comment]
Very good point!

Looks like the [build/c++11] section covers:
- Rvalue references
- Default lambda captures
- Unapproved c++11 headers
- Unapproved c++11 class/function (alignment_of, aligned_union)

Which are all probably things that might be used in the future.

Should I add the flag to stop cpplint checking c++11 in this PR or should that be done separately?

**Comment by bob-carpenter** [comment]
This one's waiting for final changes and approval from @syclik.  I was requested to review, but we don't want to double-review pull requests.

**Comment by syclik** [comment]
@andrjohns, sorry I missed that comment almost 3 weeks ago. I think it's fine here, but for future pull requests, it'd be great to have separate pull requests. No need to change the for this one. 

I'm looking at it again now.

**Pull request merged**